diff --git a/lib/base64.h b/lib/base64.h
index 9cd0183b8..a2aaa2d4a 100644
--- a/lib/base64.h
+++ b/lib/base64.h
@@ -18,11 +18,16 @@
 #ifndef BASE64_H
 # define BASE64_H
 
+/* Get _GL_ATTRIBUTE_CONST */
+# include <config-util.h>
+
 /* Get size_t. */
 # include <stddef.h>
 
-/* Get bool. */
-# include <stdbool.h>
+#ifndef GRUB_POSIX_BOOL_DEFINED
+typedef enum { false = 0, true = 1 } bool;
+#define GRUB_POSIX_BOOL_DEFINED 1
+#endif
 
 # ifdef __cplusplus
 extern "C" {
